### PR TITLE
Implemented a voting system for templates on the feed

### PR DESCRIPTION
**Pull request by bryanansong**
## Description
In this pull request I implemented a voting system for workout templates. On the frontend, I updated the FeedListItem component to include upvote and downvote buttons that make a fetch request to execute a vote. The backend now includes a route and controller method to handle vote requests. The voting system provides immediate feedback on vote actions through local changes on the frontend and correctly displays users' existing votes on the initial load.
<what this pull request is doing>
<what will be done in later PRs and not included here>

## Test Plan
image
<insert images or gifs of feature>
<otherwise, say how code was tested if not UI facing>
<any edge cases you might have specifically tested>

### CODE REVIEW

File: backend/controllers/workoutTemplatesController.js [hunk 1]
```diff
@@ -158,9 +159,81 @@ const getFeed = async (req, res) => {
...
+        userId_workoutTemplateId: {
+          userId,
+          workoutTemplateId: parseInt(templateId),
+        },
+      },
+    });
+
+    let updatedVote;
+    if (existingVote) {
+      updatedVote = await prisma.vote.update({
```
Comment (cherieliu) [inline backend/controllers/workoutTemplatesController.js:189]: should we check if the existing vote type is equal to the updated vote type? this might help save some repetitive updates to your database

**Review by cherieliu — COMMENTED** [review]
overall lgtm! very creative way to handle the up/down voting UI

File: backend/controllers/workoutTemplatesController.js (hunk 1 above)
Comment (bryanansong) [reply backend/controllers/workoutTemplatesController.js:189]: I think you're right, I will add an additional condition for
``` if ( existingVote && existingVote.voteType !== voteType )```

File: frontend/src/components/FeedListItem/FeedListItem.tsx [hunk 2]
```diff
@@ -1,10 +1,63 @@
...
+  const [upvote, setUpvote] = useState<boolean | null>(null);
+  const [netVotes, setNetVotes] = useState<number>(0);
+
+  const calculateNetVotes = () => {
+    if (!post || !post.votes) {
+      setNetVotes(0);
+      return;
+    }
+
+    const upvotes = post.votes.filter(
```
Comment (m5matthew) [inline frontend/src/components/FeedListItem/FeedListItem.tsx:18]: Could we compute this in the backend and enable an API to fetch template statistics? That way you can reuse it when building out your ranking algorithm.

File: frontend/src/components/FeedListItem/FeedListItem.tsx [hunk 3]
```diff
@@ -1,10 +1,63 @@
...
 import styles from "./FeedListItem.module.css";
 
 type FeedListItemProps = {
-  post: WorkoutTemplate,
+  post: WorkoutTemplate;
 };
 
 const FeedListItem: React.FC<FeedListItemProps> = ({ post }) => {
+  const [upvote, setUpvote] = useState<boolean | null>(null);
+  const [netVotes, setNetVotes] = useState<number>(0);
```
Comment (Wang96) [inline frontend/src/components/FeedListItem/FeedListItem.tsx:10]: I am wondering if you plan to use netVotes on client or not? I guess this will be used to rank the feed, then it might be more convenient if you calculate this and save it in backend, so that when you render the feed, you don't need to do this computation in real time.

File: frontend/src/components/FeedListItem/FeedListItem.tsx (hunk 2 above)
Comment (bryanansong) [reply frontend/src/components/FeedListItem/FeedListItem.tsx:18]: Yes that's a great idea. I just implemented this in PR 29

File: frontend/src/components/FeedListItem/FeedListItem.tsx (hunk 3 above)
Comment (bryanansong) [reply frontend/src/components/FeedListItem/FeedListItem.tsx:10]: Yeah, I think that's a great idea to help me stop repeating code. I already moved the logic for updating the netVotes to the backend. So that partly addresses this. From what I'm understanding, you would want me to create another field in the workout Template model and update this value each time a vote action is made or are the changes I made in PR 29 fine.

Comment (Wang96) [reply frontend/src/components/FeedListItem/FeedListItem.tsx:10]: Actually, I think since this is just simple calculation and everything is already inside in the votes field, it's probably fine to not save it but calculate it when you need the ranking.
